lib/integrations/awsoidc/deployservice_iam_jointoken.go

```diff
@@ -0,0 +1,199 @@
+	}.String()
+
+	return nil
+}
+
+// upsertIAMJoinToken creates or updates a Provision Token with a rule that allows Teleport Services with given AWS Identities to join the cluster.
+// The allowed Identities are the ones provided in the request (Account, Region and AWSRole).
+// It gives access for those identities to join as system roles.
+func upsertIAMJoinToken(ctx context.Context, req upsertIAMJoinTokenRequest, clt GetUpsertToken) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	rule := &types.TokenRule{
+		AWSAccount: req.accountID,
+		AWSARN:     req.awsARN,
+	}
+
+	token, err := clt.GetToken(ctx, req.tokenName)
+	switch {
+	case trace.IsNotFound(err):
+		token, err = newIAMJoinTokenWithRule(ctx, req, rule)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	case err != nil:
+		return trace.Wrap(err)
+
+	default:
+		token, err = updateTokenIAMJoin(ctx, req, rule, token)
+		if trace.IsAlreadyExists(err) {
+			// Early return when the required rule already exists.
+			return nil
+		}
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if err := clt.UpsertToken(ctx, token); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func updateTokenIAMJoin(ctx context.Context, req upsertIAMJoinTokenRequest, rule *types.TokenRule, existingToken types.ProvisionToken) (types.ProvisionToken, error) {
+	if existingToken.GetJoinMethod() != types.JoinMethodIAM {
+		return nil, trace.BadParameter("Token %q already exists but has the wrong method %q. "+
+			"Please remove it before continuing.",
+			req.tokenName, existingToken.GetJoinMethod(),
+		)
+	}
+
+	allRoles := append(existingToken.GetRoles(), types.RoleDatabase)
+	uniqueRoles := utils.Deduplicate(allRoles)
+	existingToken.SetRoles(uniqueRoles)
+
+	for _, rule := range existingToken.GetAllowRules() {
+		if rule.AWSAccount == req.accountID &&
+			rule.AWSARN == req.awsARN {
+
+			return nil, trace.AlreadyExists("required rule already exists")
+		}
+	}
+
+	existingToken.SetAllowRules(append(existingToken.GetAllowRules(), rule))
+
+	return existingToken, nil
+}
+
+// newIAMJoinTokenWithRule creates a new ProvisionToken using the
+func newIAMJoinTokenWithRule(ctx context.Context, req upsertIAMJoinTokenRequest, rule *types.TokenRule) (types.ProvisionToken, error) {
+	var systemRoles types.SystemRoles
+
+	switch req.deploymentMode {
+	case DatabaseServiceDeploymentMode:
+		systemRoles = types.SystemRoles{types.RoleDatabase}
+	default:
+		return nil, trace.BadParameter("invalid deployment mode %q, supported modes: %v", req.deploymentMode, DeploymentModes)
+	}
+
+	token := &types.ProvisionTokenV2{
+		Metadata: types.Metadata{
+			Name: req.tokenName,
+		},
+		Spec: types.ProvisionTokenSpecV2{
+			JoinMethod: types.JoinMethodIAM,
+			Roles:      systemRoles,
+			Allow:      []*types.TokenRule{rule},
+		},
+	}
+	if err := token.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return token, nil
+}
```

Comment:
Can we merge the new and update token functions into the same one? 
It will be easier to extend in the future and it's easier to understand what happens anw.

```suggestion

// upsertIAMJoinToken creates or updates a Provision Token with a rule that allows Teleport Services with given AWS Identities to join the cluster.
// The allowed Identities are the ones provided in the request (Account, Region and AWSRole).
// It gives access for those identities to join as system roles.
func upsertIAMJoinToken(ctx context.Context, req upsertIAMJoinTokenRequest, clt GetUpsertToken) error {
	if err := req.CheckAndSetDefaults(); err != nil {
		return trace.Wrap(err)
	}

	rule := &types.TokenRule{
		AWSAccount: req.accountID,
		AWSARN:     req.awsARN,
	}

	token, err := clt.GetToken(ctx, req.tokenName)
	switch {
	case err != nil && !trace.IsNotFound(err):
		return trace.Wrap(err)
	case trace.IsNotFound(err):
		token = &types.ProvisionTokenV2{}
		fallthrough
	default:
		existingToken, ok := token.(*types.ProvisionTokenV2)
		if !ok {
			return trace.BadParameter("expected token to be of type ProvisionTokenV2, got %T", token)
		}
		token, err = updateTokenIAMJoin(ctx, req, rule, existingToken)
		if trace.IsAlreadyExists(err) {
			// Early return when the required rule already exists.
			return nil
		}
		if err != nil {
			return trace.Wrap(err)
		}
	}

	if err := clt.UpsertToken(ctx, token); err != nil {
		return trace.Wrap(err)
	}

	return nil
}

func updateTokenIAMJoin(ctx context.Context, req upsertIAMJoinTokenRequest, rule *types.TokenRule, existingToken *types.ProvisionTokenV2) (types.ProvisionToken, error) {
	if existingToken.GetJoinMethod() != "" && existingToken.GetJoinMethod() != types.JoinMethodIAM {
		return nil, trace.BadParameter("Token %q already exists but has the wrong method %q. "+
			"Please remove it before continuing.",
			req.tokenName, existingToken.GetJoinMethod(),
		)
	}
	existingToken.Metadata.Name = req.tokenName
	existingToken.Spec.JoinMethod = types.JoinMethodIAM
	allRoles := existingToken.GetRoles()
	switch req.deploymentMode {
	case DatabaseServiceDeploymentMode:
		allRoles = append(allRoles, types.RoleDatabase)
	default:
		return nil, trace.BadParameter("invalid deployment mode %q, supported modes: %v", req.deploymentMode, DeploymentModes)
	}
	uniqueRoles := utils.Deduplicate(allRoles)
	existingToken.SetRoles(uniqueRoles)

	found := false
	for _, rule := range existingToken.GetAllowRules() {
		if rule.AWSAccount == req.accountID &&
			rule.AWSARN == req.awsARN {
			found = true
			break
		}
	}
	if !found {
		existingToken.SetAllowRules(append(existingToken.GetAllowRules(), rule))
	}
	return existingToken, nil
}
```